File: hard/23_merge_k_sorted_lists.rb
# 23. Merge k Sorted Lists
# Hard
#
# You are given an array of k linked-lists lists, each linked-list is sorted
# in ascending order.
#
# Merge all the linked-lists into one sorted linked-list and return it.

def merge_k_lists(lists)
  lists.filter! { |list| !list.nil? }

  a = []
  lists.each do |list|
    a.concat(to_array(list))
  end

  to_list(a.sort!)
end

def to_array(head)
  res = []

  until head.nil?
    res << head.val
    head = head.next
  end

  res
end

def to_list(arr)
  return nil if arr.empty?

  head = ListNode.new(arr[0])
  cur = head

  arr[1..].each do |a|
    cur.next = ListNode.new(a)
    cur = cur.next
  end

  head
end

File: python/23_merge_k_sorted_lists.py
"""
23. Merge k Sorted Lists
Hard

You are given an array of k linked-lists lists, each linked-list is sorted in
ascending order.

Merge all the linked-lists into one sorted linked-list and return it.
"""


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        res = []
        for l in lists:
            res.extend(self.to_array(l))

        res.sort()

        return self.to_list(res)

    def to_array(self, head: ListNode) -> List[int]:
        res = []
        while head is not None:
            res.append(head.val)
            head = head.next

        return res

    def to_list(self, list: List[int]) -> Optional[ListNode]:
        if len(list) == 0:
            return None

        head = ListNode(list[0])
        curr = head
        for v in list[1:]:
            curr.next = ListNode(v)
            curr = curr.next

        return head


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
